### PR TITLE
feat: add support for mix.lock files for elixir language

### PR DESCRIPTION
**Pull request by DmitriyLewen**
## Description
add support for mix.lock files

## Issues
- [ ] Close https://github.com/aquasecurity/trivy/issues/3346

## Related PRs
- [x] aquasecurity/trivy-db/pull/270
- [x] aquasecurity/go-dep-parser/pull/158

## Checklist
- [x] I've read the [guidelines for contributing](https://aquasecurity.github.io/trivy/latest/community/contribute/pr/) to this repository.
- [x] I've followed the [conventions](https://aquasecurity.github.io/trivy/latest/community/contribute/pr/#title) in the PR title.
- [x] I've added tests that prove my fix is effective or that my feature works.
- [ ] I've updated the [documentation](https://github.com/aquasecurity/trivy/blob/main/docs) with the relevant information (if needed).
- [ ] I've added usage information (if the PR introduces new options)
- [ ] I've included a "before" and "after" example to the description (if the PR is a user interface change).

### CODE REVIEW

File: pkg/fanal/analyzer/const.go [hunk 1]
```diff
@@ -74,6 +74,9 @@ const (
 	// C/C++
 	TypeConanLock Type = "conan-lock"
 
+	// Elixir
+	TypeElixirLock Type = "elixir-lock"
```
Comment (knqyf263) [inline pkg/fanal/analyzer/const.go:78]: I think it is Mix rather than Elixir.

File: pkg/fanal/types/const.go [hunk 2]
```diff
@@ -30,6 +30,7 @@ const (
 	RustBinary = "rustbinary"
 	Conan      = "conan"
 	Cocoapods  = "cocoapods"
+	ElixirLock = "elixir-lock"
```
Comment (knqyf263) [inline pkg/fanal/types/const.go:33]: Mix or Hex may be better. We can say we parse mix.lock and get Hex packages. Or, we parse mix.lock and get Mix dependencies.

File: pkg/fanal/analyzer/const.go (hunk 1 above)
Comment (DmitriyLewen) [reply pkg/fanal/analyzer/const.go]: Done

File: pkg/fanal/types/const.go (hunk 2 above)
Comment (DmitriyLewen) [reply pkg/fanal/types/const.go]: Done

File: pkg/fanal/analyzer/language/elixir/mixlock/mixlock.go [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+	"golang.org/x/xerrors"
+)
+
+func init() {
+	analyzer.RegisterAnalyzer(&mixLockAnalyzer{})
+}
+
+const (
+	version        = 1
+	fileNameSuffix = "mix.lock"
```
Comment (knqyf263) [inline pkg/fanal/analyzer/language/elixir/mixlock/mixlock.go:23]: Can it be a suffix? Does it mean `myproject-mix.lock` or something like that can be used?

Comment (knqyf263) [reply pkg/fanal/analyzer/language/elixir/mixlock/mixlock.go:23]: Also, it is better to define the file name here.
https://github.com/aquasecurity/trivy/blob/55cb4ad2f158beec9a5f9dd92ef004ad24d04165/pkg/fanal/types/const.go#L55-L76

File: pkg/fanal/analyzer/language/elixir/mixlock/mixlock_test.go [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+	tests := []struct {
+		name     string
+		filePath string
+		want     bool
+	}{
+		{
+			name:     "happy path",
+			filePath: "mix.lock",
+			want:     true,
+		},
```
Comment (knqyf263) [inline pkg/fanal/analyzer/language/elixir/mixlock/mixlock_test.go:75]: If Mix accepts `*mix.lock`, we should add such a test as well.

File: pkg/detector/library/driver.go [hunk 5]
```diff
@@ -51,6 +51,9 @@ func NewDriver(libType string) (Driver, error) {
 	case ftypes.Pipenv, ftypes.Poetry, ftypes.Pip, ftypes.PythonPkg:
 		ecosystem = vulnerability.Pip
 		comparer = pep440.Comparer{}
+	case ftypes.Hex:
+		ecosystem = vulnerability.Erlang
+		comparer = compare.GenericComparer{}
```
Comment (knqyf263) [inline pkg/detector/library/driver.go:56]: Which version spec Hex uses? [Semantic versioning](https://semver.org/) is widely used, but some languages use different spec like [PEP440](https://peps.python.org/pep-0440/).

Comment (DmitriyLewen) [reply pkg/detector/library/driver.go:56]: Hex uses semver 2.0 - https://hexdocs.pm/elixir/Version.html

Comment (knqyf263) [reply pkg/detector/library/driver.go:56]: This document is about Elixir modules, isn't it? Do Hex packages also follow semver?

Comment (DmitriyLewen) [reply pkg/detector/library/driver.go:56]: i found info about publish new hex packages.
It says "First, make sure that the :version property is correct. All Hex packages are required to follow [semantic versioning](http://semver.org/)."
https://hex.pm/docs/publish#adding-metadata-to-code-classinlinemixexscode

Comment (DmitriyLewen) [reply pkg/detector/library/driver.go:56]: There is also `version module` information in the [hex documents](https://hex.pm/docs/usage#defining-dependencies). But the link to the elixir modules (which I posted earlier)

File: pkg/fanal/analyzer/language/elixir/mixlock/mixlock.go (hunk 3 above)
Comment (DmitriyLewen) [reply pkg/fanal/analyzer/language/elixir/mixlock/mixlock.go]: you are right. Thanks. I moved this to `trivy/pkg/fanal/types/const.go`

File: pkg/fanal/analyzer/language/elixir/mixlock/mixlock_test.go (hunk 4 above)
Comment (DmitriyLewen) [reply pkg/fanal/analyzer/language/elixir/mixlock/mixlock_test.go:76]: looks like lock file doesn't have rules for name. Name may have any name.
https://hexdocs.pm/mix/Mix.Project.html#module-configuration.

I added footnote on docs for mix(as for conan).
https://github.com/aquasecurity/trivy/blob/afd32a36284e8a90371ff6cf429767ee6a68ef49/docs/docs/vulnerability/detection/language.md?plain=1#L30

I think this is better way.

File: pkg/detector/library/driver.go (hunk 5 above)
Comment (knqyf263) [reply pkg/detector/library/driver.go:56]: Great! We can use the semver library.

File: pkg/fanal/analyzer/language/elixir/mixlock/mixlock.go [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+func (a mixLockAnalyzer) Analyze(_ context.Context, input analyzer.AnalysisInput) (*analyzer.AnalysisResult, error) {
+	p := mix.NewParser()
+	res, err := language.Analyze(types.Hex, input.FilePath, input.Content, p)
+	if err != nil {
+		return nil, xerrors.Errorf("%s parse error: %w", input.FilePath, err)
+	}
+	return res, nil
+}
+
+func (a mixLockAnalyzer) Required(_ string, fileInfo os.FileInfo) bool {
```
Comment (knqyf263) [inline pkg/fanal/analyzer/language/elixir/mixlock/mixlock.go:36]: I prefer using `filePath` so tests can be simple.

File: pkg/fanal/analyzer/language/elixir/mixlock/mixlock_test.go (hunk 4 above)
Comment (knqyf263) [reply pkg/fanal/analyzer/language/elixir/mixlock/mixlock_test.go:76]: We usually support the default name only. If people want to use other names, they can specify `--file-pattern`. It makes sense to support only `mix.lock`.

>:lockfile - the name of the lockfile used by the mix deps.* family of tasks. Defaults to "mix.lock".

**Review by knqyf263 — APPROVED** [review]

File: pkg/fanal/analyzer/language/elixir/mixlock/mixlock.go (hunk 6 above)
Comment (DmitriyLewen) [reply pkg/fanal/analyzer/language/elixir/mixlock/mixlock.go]: Okay. Thank you for changing this!

**Pull request merged**